Fix EuiComboBox focus trap

cjcenizal opened the pull request:
Alternative approach to https://github.com/elastic/eui/pull/845. Fixes https://github.com/elastic/eui/issues/788 and https://github.com/elastic/eui/issues/787. Also fixes a bug in which clicking the caret doesn't set focus to the search input.

@nreese @chandlerprall What do you think of this approach? If you think this has promise then I need to update the CHANGELOG with the many breaking changes, update other instances in which `EuiFormControlLayout` is consumed, and fix and flesh out the tests.

![combo_box_focus_trap](https://user-images.githubusercontent.com/1238659/40400691-280f6172-5df8-11e8-989d-5d9e14e9f894.gif)

cjcenizal commented on src/components/form/form_control_layout/form_control_layout.js:
```diff
@@ -99,18 +125,24 @@ export const EuiFormControlLayout = ({
-  icon: PropTypes.string,
+  icon: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.shape({
```
Thoughts on this kind of pattern? I was inspired a bit by https://www.youtube.com/watch?v=Yy7gFgETp0o&t=1079s.

chandlerprall replied:
I like it

chandlerprall commented:
Do we know what browser/OS combinations do or don't allow tabbing out of a native `select` box? Chrome/Mac at least does not let you tab out of the expanded options list.

cjcenizal commented:
On Mac:

* Safari blocks you from tabbing out of the expanded options list.
* Firefox will select the option you've highlighted and close the list.

chandlerprall commented:
This approach does look a lot cleaner to me than #845

nreese commented:
> This approach does look a lot cleaner to me than #845

agreed. I will close 845 in favor of this

cjcenizal commented on src/components/combo_box/combo_box.js:
```diff
@@ -450,8 +463,14 @@ export class EuiComboBox extends Component {
   componentDidMount() {
     this._isMounted = true;
+    document.addEventListener('click', this.onDocumentFocusChange);
+    document.addEventListener('focusin', this.onDocumentFocusChange);
```
By leaving these event listeners active while the component is mounted, we can open the list when the user tabs backwards to give the clear button focus.

cjcenizal commented:
@chandlerprall @nreese I updated the tests, made some minor tweaks, and updated the CHANGELOG. Could you take another look please?

chandlerprall commented on src/components/combo_box/combo_box.js:
```diff
@@ -138,25 +138,37 @@ export class EuiComboBox extends Component {
       }
+
+      // Otherwise tab to the next adjacent item.
+      tabbableItems[searchInputIndex + amount].focus();
```
test if this exceeds `tabbableItems.length`?

chandlerprall commented on src/components/combo_box/combo_box.js:
```diff
@@ -138,25 +138,37 @@ export class EuiComboBox extends Component {
+
+      // Wrap to last tabbable if tabbing backwards.
+      if (amount < 0) {
+        if (searchInputIndex === 0) {
```
`amount` may be less that `-1`, test if `searchInputIndex + amount < 0`

cjcenizal commented:
@chandlerprall Thanks for the review! I addressed your comments by adding a check that the `amount` can only be -1 or 1 (which seems reasonable since this should only accept valid tabbing increments). This also solves issues where we tab beyond `tabbableItems.length`, since the other conditions in this method will wrap focus to the beginning of the tab order. Could you take another look?

chandlerprall commented on src/components/combo_box/combo_box.js:
```diff
@@ -137,26 +137,42 @@ export class EuiComboBox extends Component {
   };
 
   tabAway = amount => {
+    if (![-1, 1].includes(amount)) {
```
IE doesn't support `includes`. What do you think of `if (Math.abs(amount) !== 1)` ?

cjcenizal replied:
Oh interesting! We're using `includes` in other places in EUI. Maybe we should just add documentation stating that we expect consumers to polyfill ES2015 features, e.g. with [babel-polyfill](https://babeljs.io/docs/usage/polyfill/). I think that's fair, since EUI is intended for usage within Elastic and I'd be surprised if we couldn't meet that requirements.

chandlerprall replied:
Sounds like the right thing since the methods are already being used

cjcenizal commented:
I had to refactor `EuiFormControlLayout` significantly because it wasn't placing the DOM elements in the correct order when the icon is on the right with the clear button. @snide @cchaos Could you take a look at these layouts and make sure they make sense from a design perspective?

![image](https://user-images.githubusercontent.com/1238659/40517440-2e972188-5f6a-11e8-9024-8d2956fd857d.png)

snide commented on src/components/form/field_number/field_number.js:
```diff
@@ -81,7 +82,14 @@ EuiFieldNumber.propTypes = {
   step: PropTypes.number,
   value: numberOrEmptyString,
-  icon: PropTypes.string,
+  icon: PropTypes.oneOfType([
```
A lot of these components purposefully limited prop declarations. Is there a reason to add iconSide and onClick to the number field? I kind of like giving people less options here which is why I only exposed the props in some of the compoents. Generally, when we've given people the power to use things, they start using them. I'm just a little worried about seeing random action icons on the right side of inputs that lack context. Down arrow, clear? Makes sense. Random logstash icon with some hidden meaning and an onClick action has me worried 😄

snide replied:
Yep. Looks like using them will actually break the layouts because of the assumptions the previous api made.

I think if these just get reverted to the strings only you're prolly ok? Leave the mapping stuff to the control only?

![image](https://user-images.githubusercontent.com/324519/40526871-1ccdc842-5f9e-11e8-9f13-cf433bd65c97.png)

cjcenizal replied:
Oh great catch! This was an oversight on my part.

snide commented on src/components/form/form_control_layout/_form_control_layout.scss:
```diff
@@ -13,80 +13,49 @@
-    // Loading spinner needs adjustment if clear also exists
-    ~ .euiFormControlLayout__loading {
-      right: $euiFormControlPadding*3;
+    > * + * {
```
Looking at this again, it's only a problem in the generic examples and not the real form controls themselves. So ignore the above. But we might want to provide a comment about the issue if someone wants to use this stuff.

snide replied:
^^ Well, kinda. See the comment below about the number field.

cjcenizal replied:
Good idea, I'll add a note to the docs.

cjcenizal commented:
@snide Thanks for your feedback! I've addressed your points. About the pattern of an `icon` prop accepting an object, I think we'll have to consider it on a case-by-case basis. It's good in situations where the icon needs so much customization like inside of `combo_box.js`, but makes much less sense for something a button's icon, where these extra props won't do much good. I think the real benefit here is that we have a new pattern to apply to props which require this extra level of customization, which we can reach for when needed.

nreese commented:
works great

chandlerprall commented:
Thanks for pushing forward on this! LGTM
